File: web-admin/src/types/bus.ts
/**
 * Types pour les bus et données GPS
 */

export type BusStatus = 'EN_ROUTE' | 'EN_RETARD' | 'A_L_ARRET' | 'HORS_SERVICE';

export interface GPSPosition {
  lat: number;
  lng: number;
  speed: number;
  timestamp: number;
}

export interface GPSHistoryPoint {
  location: string;
  time: string;
  lat: number;
  lng: number;
}

export interface Incident {
  type: 'pneu_creve' | 'retard_imprevu' | 'porte_bloquee' | 'autre';
  label: string;
  date: string;
  resolved: boolean;
}

export interface MaintenanceRecord {
  type: 'vidange' | 'controle_technique' | 'reparation' | 'autre';
  label: string;
  date: string;
  status: 'scheduled' | 'approved' | 'completed' | 'pending';
  statusLabel: string;
}

export interface Bus {
  id: string;
  number: string; // Format: BUS-XX (ex: BUS-12, BUS-45)
  immatriculation: string;
  chauffeur: string;
  chauffeurId?: string;
  capacite: number;
  itineraire: string;
  photoUrl?: string;
  status: BusStatus;
  statusLabel: string;
  currentPosition?: GPSPosition;
  lastGPSUpdate?: string;
  gpsHistory?: GPSHistoryPoint[];
  incidents?: Incident[];
  maintenanceStatus: number; // Pourcentage 0-100
  maintenanceRecords?: MaintenanceRecord[];
}

export interface DashboardStats {
  // Métriques de base
  busActifs: number;
  busTotaux: number;
  elevesTransportes: number;
  busEnRetard: number;
  totalTrajets: number;
  alertesMaintenance: number;
  
  // Nouvelles métriques opérationnelles (contexte Abidjan)
  retardsCritiques: number; // Retards > 15 minutes
  retardsGraves: number; // Retards > 20 minutes
  busImmobilises: number; // Bus en panne/hors service
  busDisponibles: number; // Bus disponibles (totaux - immobilisés)
  tauxValidation: number; // Pourcentage élèves scannés (0-100)
  elevesNonScannes: number; // Nombre élèves non scannés
  
  // État détaillé du service
  busEnRoute: number; // Bus actuellement en déplacement
  busArrives: number; // Bus arrivés à destination
  busNonPartis: number; // Bus n'ayant pas encore démarré
  busEnAttente: number; // Bus à l'arrêt (attente élèves)
  
  // Métriques de performance (trafic vs prévision)
  retardMoyen: number; // Retard moyen en minutes
  tauxPonctualite: number; // % de bus à l'heure (0-100)
  tempsTrajetMoyen: number; // Temps de trajet moyen (minutes)
  tempsTrajetPrevu: number; // Temps prévu initialement
}

File: mobile-parent/parent_app/lib_backup/models/bus.dart
/// Modèle de données pour un bus
class Bus {
  final String id;
  final String immatriculation;
  final String chauffeur;
  final String chauffeurId;
  final int capacite;
  final String itineraire;
  final BusStatus status;
  final String statusLabel;
  final GPSPosition? currentPosition;
  final String? lastGPSUpdate;
  final int maintenanceStatus;

  Bus({
    required this.id,
    required this.immatriculation,
    required this.chauffeur,
    required this.chauffeurId,
    required this.capacite,
    required this.itineraire,
    required this.status,
    required this.statusLabel,
    this.currentPosition,
    this.lastGPSUpdate,
    required this.maintenanceStatus,
  });

  factory Bus.fromJson(Map<String, dynamic> json) {
    return Bus(
      id: json['id'] as String,
      immatriculation: json['immatriculation'] as String,
      chauffeur: json['chauffeur'] as String,
      chauffeurId: json['chauffeurId'] as String,
      capacite: json['capacite'] as int,
      itineraire: json['itineraire'] as String,
      status: _parseBusStatus(json['status'] as String),
      statusLabel: json['statusLabel'] as String,
      currentPosition: json['currentPosition'] != null
          ? GPSPosition.fromJson(json['currentPosition'])
          : null,
      lastGPSUpdate: json['lastGPSUpdate'] as String?,
      maintenanceStatus: json['maintenanceStatus'] as int,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'immatriculation': immatriculation,
      'chauffeur': chauffeur,
      'chauffeurId': chauffeurId,
      'capacite': capacite,
      'itineraire': itineraire,
      'status': status.name,
      'statusLabel': statusLabel,
      'currentPosition': currentPosition?.toJson(),
      'lastGPSUpdate': lastGPSUpdate,
      'maintenanceStatus': maintenanceStatus,
    };
  }

  static BusStatus _parseBusStatus(String status) {
    switch (status) {
      case 'EN_ROUTE':
        return BusStatus.enRoute;
      case 'EN_RETARD':
        return BusStatus.enRetard;
      case 'A_L_ARRET':
        return BusStatus.aLArret;
      case 'HORS_SERVICE':
        return BusStatus.horsService;
      default:
        return BusStatus.enRoute;
    }
  }
}

/// Statut du bus
enum BusStatus {
  enRoute,
  enRetard,
  aLArret,
  horsService,
}

/// Position GPS
class GPSPosition {
  final double lat;
  final double lng;
  final double speed;
  final int timestamp;

  GPSPosition({
    required this.lat,
    required this.lng,
    required this.speed,
    required this.timestamp,
  });

  factory GPSPosition.fromJson(Map<String, dynamic> json) {
    return GPSPosition(
      lat: (json['lat'] as num).toDouble(),
      lng: (json['lng'] as num).toDouble(),
      speed: (json['speed'] as num).toDouble(),
      timestamp: json['timestamp'] as int,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'lat': lat,
      'lng': lng,
      'speed': speed,
      'timestamp': timestamp,
    };
  }
}

